### PR TITLE
Open Enterprise Agent

### PR DESCRIPTION
**Pull request by BS-IO**
The Open Enterprise Agent is a software agent for building decentralised identity applications and services in line with W3C standards and Hyperledger Aries RFCs. Aiming to be blockchain agnostic and interoperable across two key technology ecosystems in SSI, it provides a scalable, robust, performant and user-friendly platform for building solutions and products based on it. Its REST API enables easy integration with any programming language, on top of a core built in Scala, a language well suited for building fault-tolerant, highly concurrent systems.

### CODE REVIEW

**Comment by darrellodonnell** [comment]
Fully supported. I do want to see how this maps to other Aries agents (e.g. ACA-Py, AFJ), and we should see where there is/isn't overlap.

**Comment by 2byrds** [comment]
RootsID is in full support of this proposal. The open enterprise agent is ahead of most others in terms of DIDComm v2, Scala will be familiar/perform well for JVM enterprise users, did:prism is a full-featuted DID method, and the OEA support of the AnonCreds transition has been great for the HL community.

**Comment by TimoGlastra** [comment]
I'm fully supportive of this new initiative and look forward to using the Open Enterprise Agent

File: labs/open-enterprise-agent.md
Comment (tkuhrt) [inline labs/open-enterprise-agent.md:81]: How are the initial maintainers different from the initial committers?

File: labs/open-enterprise-agent.md
Comment (tkuhrt) [inline labs/open-enterprise-agent.md:97]: I could not look at this repo. Is it public? I would like to understand the license and ensure DCO sign-off of all commits in this repo, which is a standard step that I prior to approving a lab.

File: labs/open-enterprise-agent.md
Comment (tkuhrt) [inline labs/open-enterprise-agent.md:98]: I am not sure that I understand this comment. How does the use of the library impact the license of the repo that you are looking to bring over?

File: labs/open-enterprise-agent.md
Comment (BS-IO) [reply labs/open-enterprise-agent.md:97]: It's not public at present, I will work with our team to make that happen if needed.

File: labs/open-enterprise-agent.md
Comment (tkuhrt) [reply labs/open-enterprise-agent.md:81]: Will these two sets have different permissions?

Comment (BS-IO) [reply labs/open-enterprise-agent.md:81]: Yes. Once the code-base is transitioned, we intend on moving towards a community-based governance model, and would like to ensure that the maintainer list ensure a balanced representation of the community we intend to serve. In short: most developers will become 'regular' contributors.

File: labs/open-enterprise-agent.md
Comment (BS-IO) [reply labs/open-enterprise-agent.md:97]: I will work with the team to ensure you get access. Given it's near nighttime in our timezone, I hope we can get that sorted tomorrow morning.

File: labs/open-enterprise-agent.md
Comment (tkuhrt) [reply labs/open-enterprise-agent.md:81]: What I am trying to figure out is what GitHub teams will exist for this lab and what [GitHub permissions](https://docs.github.com/en/organizations/managing-user-access-to-your-organizations-repositories/repository-roles-for-an-organization) will be required? Will there be one team or two teams? If two teams what GitHub permissions will each team need?

Comment (BS-IO) [reply labs/open-enterprise-agent.md:81]: Ah got it. I'd argue that there's 2 sets of permissions, maintain (for the maintainers) and write (for contributors)

**Comment by dhh1128** [comment]
Thumbs up from me as a sponsor.

File: labs/open-enterprise-agent.md
Comment (BS-IO) [reply labs/open-enterprise-agent.md:97]: @tkuhrt you should have access to the repository. Please let me know if there's anything else you need.

**Comment by victormartinez-work** [comment]
I fully support this new initiative and the importance of having an agent capable of didcomm v2 and support of w3c credential data model VC.

File: labs/open-enterprise-agent.md
Comment (tkuhrt) [reply labs/open-enterprise-agent.md:97]: I do have access to the repository. This leads to the following questions:

1. I am confused by the LICENSES.md file. It appears that there is a separate license listed for each of the dependencies. What is the LICENSE for the source code that you are looking to bring into Hyperledger?
2. The code does not have DCO sign-off, which is required for code contributed to Hyperledger Labs. There are two ways that we can resolve this. (1) Retroactively gather DCO sign-offs for the existing commits (2) Create a new repo and commit the code without the history, but using DCO sign-off. Which approach do you plan to take with this repo?

Comment (antonbaliasnikov) [reply labs/open-enterprise-agent.md:97]: Hi @tkuhrt !

I would like to clarify about the DCO requirement:
> The code does not have DCO sign-off, which is required for code contributed to Hyperledger Labs

For Atala PRISM, we're using [commits signature verification](https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification) and we sign each commit with our PGP keys that we officially publish and associate with an e-mail of each contributor.

We have a requirement that every commit must be signed and verified with a contributor PGP key.

Does the DCO requirement to put an additional string `Signed-off-by ...` to the body of each commit apply in this case? Could you, please, clarify if it is still required if the commit has already passed a signature verification?

Thank you!

**Comment by BS-IO** [comment]
Thanks - we will discuss how we can best meet this condition on Monday with
our technical teams.

**Review by dhh1128 — COMMENTED** [review]
I would just like to say that the verbiage about me overqualifies me a bit. I *was* a member of the ACDC task force, and maybe I was once a primary editor for a few weeks. I still attend ACDC task force meetings. However, my editor status has faded into being more of a supporter of the initiative. Also, I am not currently a lead on the TSPTF; I *was* a lead for the first 2-3 months but recently stepped back. So: is the qualification supposed to be current, or historical?

**Comment by arsulegai** [comment]
I would like to get access to https://github.com/input-output-hk/atala-prism-building-blocks. In support of the proposal, looking forward to seeing how the agent proposed here is comprehending other Aries agents.

**Comment by knagware9** [comment]
I fully support this lab but also want to know why this repo-https://github.com/input-output-hk/atala-prism-building-blocks  can't be transferred to Hyperledger. Is there any dependency of this repo to code which will be open source in hyperledger-labs.

File: labs/open-enterprise-agent.md
Comment (knagware9) [reply labs/open-enterprise-agent.md:98]: I fully support this lab but also want to know why this repo-https://github.com/input-output-hk/atala-prism-building-blocks can't be transferred to Hyperledger. Is there any dependency of this repo to code which will be open source in hyperledger-labs.

**Review by knagware9 — CHANGES_REQUESTED** [review]
I want to know why this repo-https://github.com/input-output-hk/atala-prism-building-blocks can't be transferred to Hyperledger. Is there any dependency of this repo to code which will be open source in hyperledger-labs.

File: labs/open-enterprise-agent.md
Comment (nidhi-singh02) [inline labs/open-enterprise-agent.md:96]: I would like to understand how this repository is getting used in this lab. 
Could you provide me access to [this](https://github.com/input-output-hk/atala-prism-building-blocks) repository.

File: labs/open-enterprise-agent.md
Comment (nidhi-singh02) [inline labs/open-enterprise-agent.md:8]: Wanted to confirm does SSI here stands for "Self-sovereign identity". 
It would be better if you can specify it in the Description along with abbreviation.

**Comment by lohanspies** [comment]
Fully support the Open Enterprise Agent proposal.

File: labs/open-enterprise-agent.md
Comment (FayyaadhAdams) [reply labs/open-enterprise-agent.md:98]: Hey @knagware9   - I think there is a little confusion. The code that we are going to be moving to HL repo is -https://github.com/input-output-hk/atala-prism-building-blocks , the code that we will not transfer is https://github.com/qos-ch/logback.

https://github.com/input-output-hk/atala-prism-building-blocks makes use of [//github.com/qos-ch/logback](https://input-output-rnd.slack.com//github.com/qos-ch/logback) and we were advised by our legal team to include it in the proposal because it is part of the GNU Lesser General Public License (https://en.wikipedia.org/wiki/GNU_Lesser_General_Public_License). 

Our legal team has advised that we include https://input-output-rnd.slack.com//github.com/qos-ch/logback from a legal perspective in the proposal but it is a free software license under the GNU Lesser Public License so we do not own it and only make use of it in https://github.com/input-output-hk/atala-prism-building-blocks that we plan to move to the HL repo.

Hope that I was able to answer your question

**Comment by ryjones** [comment]
@hyperledger-labs/lab-stewards @FayyaadhAdams @BS-IO I propose the repo come in squashed and signed via DCO, and we reference but do not transfer the GPL repo. Thoughts?

File: labs/open-enterprise-agent.md
Comment (BS-IO) [reply labs/open-enterprise-agent.md:96]: @knagware9 Not sure if there's a misunderstanding, but the https://github.com/input-output-hk/prism-building-blocks  repo is the one we intend to transfer to Hyperledger. What we wanted to clarify is that we are making use of a library which is GNU LGPL Licensed. I don't think that's a problem, but I'm not a legal expert.

**Comment by BS-IO** [comment]
> I would just like to say that the verbiage about me overqualifies me a bit. I _was_ a member of the ACDC task force, and maybe I was once a primary editor for a few weeks. I still attend ACDC task force meetings. However, my editor status has faded into being more of a supporter of the initiative. Also, I am not currently a lead on the TSPTF; I _was_ a lead for the first 2-3 months but recently stepped back. So: is the qualification supposed to be current, or historical?

@dhh1128 I'll update it to _Former Lead of ToIP Trust Spanning Task Force, Contributor ToIP ACDC Task Force, Aries RFC Maintainer_ if that's better suited?

**Comment by BS-IO** [comment]
> @hyperledger-labs/lab-stewards @FayyaadhAdams @BS-IO I propose the repo come in squashed and signed via DCO, and we reference but do not transfer the GPL repo. Thoughts?

- Agree that we don't have to transfer the GPL repo but just reference it.

- I'd prefer not to squash as I don't feel it's good practice to get rid of the history when open sourcing, and we're using [commits signature verification](https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification), signing each commit with our PGP keys that we officially publish and associate with an e-mail of each contributor. However, if it's deemed this is not possible, then I don't see another option?

File: labs/open-enterprise-agent.md
Comment (BS-IO) [reply labs/open-enterprise-agent.md:97]: We will be working on seeing if we can resign to meet the dco requirement. We expect results to be clear next week. If that doesn't work, we'll likely squash as we move over.

File: labs/open-enterprise-agent.md
Comment (BS-IO) [reply labs/open-enterprise-agent.md]: Updated this.

File: labs/open-enterprise-agent.md
Comment (knagware9) [reply labs/open-enterprise-agent.md:96]: Okay

**Review by knagware9 — APPROVED** [review]

**Comment by ryjones** [comment]
@BS-IO 

> * I'd prefer not to squash as I don't feel it's good practice to get rid of the history when open sourcing, and we're using [commits signature verification](https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification), signing each commit with our PGP keys that we officially publish and associate with an e-mail of each contributor. What we will try and do is to see if we can retro-actively get the DCO in order (we'll likely know by early next week if this is possible). However, if this is not possible, then I don't see another option?

You can retroactively DCO sign them. [this explains the process](https://github.com/dcoapp/app/blob/main/README.md); [here is an example commit](https://github.com/hyperledger/aries-cloudagent-loadgenerator/commit/8844d8e43a3da735c390aba825d47861724c8842).

**Comment by BS-IO** [comment]
> @BS-IO 
> 
> > * I'd prefer not to squash as I don't feel it's good practice to get rid of the history when open sourcing, and we're using [commits signature verification](https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification), signing each commit with our PGP keys that we officially publish and associate with an e-mail of each contributor. What we will try and do is to see if we can retro-actively get the DCO in order (we'll likely know by early next week if this is possible). However, if this is not possible, then I don't see another option?
> 
> You can retroactively DCO sign them. [this explains the process](https://github.com/dcoapp/app/blob/main/README.md); [here is an example commit](https://github.com/hyperledger/aries-cloudagent-loadgenerator/commit/8844d8e43a3da735c390aba825d47861724c8842).
> 
> 

Thanks, we will try that. Hopefully we can sort it out by mid next week.

File: labs/open-enterprise-agent.md
Comment (nidhi-singh02) [inline labs/open-enterprise-agent.md:27]: Q: What does server mean here ? Does it mean Virtual machine or bare metal server or anything else ? Could you be more specific.

Comment (davidpoltorak-io) [reply labs/open-enterprise-agent.md:27]: Thank you for your question - and thank you for your patience with our response! 

The agent is a JVM based application (Scala) and when run in our development infrastructure - or running locally using utiltiies in the repo - is containerised using docker. We use cross-platform builds in our CI pipelines such that images are available for Linux and Mac (Intel and Apple Silicon) - I'll add a commit to this pull request to detail these and expand on this section.

**Comment by BS-IO** [comment]
We've retroactively DCO signed. https://github.com/input-output-hk/atala-prism-building-blocks/compare/a271709a789fd28d9404d5e0660b0d84bce41d1b...3c38dae89e8852ee181b69aa06910909d654a447

File: labs/open-enterprise-agent.md
Comment (BS-IO) [reply labs/open-enterprise-agent.md:105]: We've retroactively DCO signed the repo: https://github.com/input-output-hk/atala-prism-building-blocks/compare/a271709a789fd28d9404d5e0660b0d84bce41d1b...3c38dae89e8852ee181b69aa06910909d654a447.

**Review by tkuhrt — APPROVED** [review]

**Comment by tkuhrt** [comment]
@BS-IO : Can you please rebase this PR to the latest main branch? We can then merge this and get the repo's transferred.

**Comment by BS-IO** [comment]
> @BS-IO : Can you please rebase this PR to the latest main branch? We can then merge this and get the repo's transferred.

@tkuhrt done.

**Comment by tkuhrt** [comment]
@BS-IO : Please work with @ryjones to transfer the repo

**Comment by ryjones** [comment]
@BS-IO any thoughts on when to transfer the repo to labs?

**Pull request merged**
